base/linalg/triangular.jl

```diff
@@ -1369,34 +1369,217 @@ end
-    maxsqrt = 100
     theta = [1.586970738772063e-005,
          2.313807884242979e-003,
          1.938179313533253e-002,
          6.209171588994762e-002,
          1.276404810806775e-001,
          2.060962623452836e-001,
          2.879093714241194e-001]
+
+    n = checksquare(A0)
+
+    A, m, s = invsquaring(A0,theta)
+
+    # Compute accurate diagonal of A - I
+    sqrt_diag!(A0, A, s)
+
+    # Compute the Gauss-Legendre quadrature formula
+    x, w = Base.QuadGK.gauss(Float64, m)
+    for i = 1:m
+        x[i] += 1
+        x[i] *= 0.5
+    end
+    scale!(w, 0.5)
+
+    # Compute the Padé approximation
+    Y = UpperTriangular(zeros(T, n, n))
+    D = similar(A)
+    Dtmp = similar(A)
+    triu!(Dtmp)
+    for k = 1:m
+        scale!(D,A,x[k])
+        for i = 1:n
+            @inbounds D[i,i] = D[i,i] + 1
+        end
+        scale!(Dtmp,A,w[k])
+        A_rdiv_B!(Dtmp.data,D)
+        for j = 1:n
+            for i = 1:j
+                @inbounds Y[i,j] = Y[i,j] + Dtmp[i,j]
+            end
+        end
+    end
+
+    # Scale back
+    scale!(2^s,Y.data)
+
+    # Compute accurate diagonal and superdiagonal of log(A)
+    for k = 1:n-1
+        Ak = complex(A0[k,k])
+        Akp1 = complex(A0[k+1,k+1])
+        logAk = log(Ak)
+        logAkp1 = log(Akp1)
+        Y[k,k] = logAk
+        Y[k+1,k+1] = logAkp1
+        if Ak == Akp1
+            Y[k,k+1] = A0[k,k+1] / Ak
+        elseif 2 * abs(Ak) < abs(Akp1) || 2 * abs(Akp1) < abs(Ak)
+            Y[k,k+1] = A0[k,k+1] * (logAkp1 - logAk) / (Akp1 - Ak)
+        else
+            w = atanh((Akp1 - Ak)/(Akp1 + Ak) + im * pi * unw(logAkp1-logAk))
+            Y[k,k+1] = 2 * A0[k,k+1] * w / (Akp1 - Ak)
+        end
+    end
+    if isreal(Y)
+        return UpperTriangular(real(Y))
+    else
+        return UpperTriangular(Y)
+    end
+end
+logm(A::LowerTriangular) = logm(A.').'
+
+# Auxiliary functions for logm and matrix power
+
+# Compute accurate diagonal of A = A0^s - I
+#   Al-Mohy, "A more accurate Briggs method for the logarithm",
+#      Numer. Algorithms, 59, (2012), 393–402.
+function sqrt_diag!(A0::UpperTriangular, A::UpperTriangular, s)
+    n = checksquare(A0)
+    for i = 1:n
+        a = complex(A0[i,i])
+        if s == 0
+            A[i,i] = a - 1
+        else
+            s0 = s
+            if imag(a) >= 0 && real(a) <= 0 && a != 0
+                a = sqrt(a)
+                s0 = s - 1
+            end
+            z0 = a - 1
+            a = sqrt(a)
+            r = 1 + a
+            for j = 1:s0-1
+                a = sqrt(a)
+                r = r * (1 + a)
+            end
+            A[i,i] = z0 / r
+        end
+    end
+end
+
+# Repeatedly compute the square roots of A so that in the end its
+# eigenvalues are close enough to the positive real line
+function invsquaring{T}(A0::UpperTriangular{T}, theta)
+    maxsqrt = 100
     tmax = size(theta, 1)
-    n = size(A0, 1)
+    n = checksquare(A0)
     A = copy(A0)
```

Comment:
Would it make preemptively define `A` as complex: `A = complex(copy(A0))`? As later in the loop we have `A = sqrtm(A)`, it should simplify `A` type inference and potentially make the things faster.